In [2]:
import json

In [3]:
# Provide clear descriptions so Gemini understands WHEN to use them
SQL_TEMPLATES = {
    "daily_revenue": {
        "description": "Calculates the daily total sales revenue trend over a specified date range. Trigger this when the user asks for daily sales, daily revenue, or earnings day-by-day.",
        "query": """SELECT o.Order_Date as order_date, SUM(Sales) as total_sales  FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID WHERE Order_Date >= %s AND Order_Date <= %s GROUP BY o.Order_Date ORDER BY o.Order_Date DESC;"""
    },
    "current_month_revenue": {
        "description": "Calculates the total sales revenue aggregated by month over a specified date range. Trigger this when the user asks for monthly revenue trends, monthly sales, or performance month-by-month.",
        "query": """SELECT MONTH(o.Order_Date) as order_month, SUM(Sales) as total_sales FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID WHERE Order_Date >= %s AND Order_Date <= %s GROUP BY MONTH(o.Order_Date) ORDER BY MONTH(o.Order_Date) DESC;"""
    },
    "top_category": {
        "description": "Identifies the highest revenue-generating product categories over a specific date range. Trigger this when the user asks which broad categories are selling the most or generating the highest revenue.",
        "query":"""SELECT p.Category as category, SUM(Sales) as total_sales FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID INNER JOIN products AS p ON p.Product_ID = d.Product_ID WHERE Order_Date >= %s AND Order_Date <= %s GROUP BY p.Category ORDER BY SUM(Sales) DESC;"""
    },
    "top_subcategory": {
        "description": "Identifies the highest revenue-generating product sub-categories over a specific date range. Trigger this for a detailed breakdown of best-selling sub-categories or specific item groups.",
        "query": """SELECT p.Sub_Category as sub_category, SUM(Sales) as total_sales FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID INNER JOIN products AS p ON p.Product_ID = d.Product_ID WHERE Order_Date >= %s AND Order_Date <= %s GROUP BY p.Sub_Category ORDER BY SUM(Sales) DESC;"""
    },
    "top_profit": {
        "description": "Calculates the total net profit aggregated by day over a specific date range. Trigger this when the user asks about profitability, net earnings after costs, or the most profitable days.",
        "query": """SELECT o.Order_Date as order_date, SUM(Profit) as total_profit FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID WHERE Order_Date >= %s AND Order_Date <= %s GROUP BY o.Order_Date ORDER BY SUM(Profit) DESC;"""
    },
    "top_quantity_product": {
        "description": "Identifies the specific products with the highest volume of units sold over a specific date range. Trigger this when the user asks for best-selling items by physical quantity or inventory movement.",
        "query": """SELECT p.Product_Name as product_name, SUM(quantity) as total_quantity FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID INNER JOIN products AS p ON p.Product_ID = d.Product_ID WHERE Order_Date >= %s  AND Order_Date <= %s  GROUP BY p.Product_Name ORDER BY SUM(quantity) DESC;"""
    },
    "top_discount": {
        "description": "Analyzes the frequency of different discount rates applied to orders over a specific date range. Trigger this when the user asks about promotional performance, marketing campaigns, or how often discounts are used.",
        "query": """SELECT d.discount as discount, COUNT(*) as number_of_discount FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID WHERE Order_Date >= %s AND Order_Date <= %s AND d.discount <> 0 GROUP BY d.discount ORDER BY COUNT(*) DESC;"""
    },
}

In [19]:
sql_query=SQL_TEMPLATES.get("daily_revenue").get("query")

parts = sql_query.split(" GROUP BY ")

part_1 = parts[0]
part_2 = " GROUP BY " + parts[1]

print(parts)
print("Phần 1:", part_1)
print("Phần 2:", part_2)

['SELECT o.Order_Date as order_date, SUM(Sales) as total_sales  FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID WHERE Order_Date >= %s AND Order_Date <= %s', 'o.Order_Date ORDER BY o.Order_Date DESC;']
Phần 1: SELECT o.Order_Date as order_date, SUM(Sales) as total_sales  FROM orders AS o INNER JOIN order_details AS d ON o.Order_ID = d.Order_ID WHERE Order_Date >= %s AND Order_Date <= %s
Phần 2:  GROUP BY o.Order_Date ORDER BY o.Order_Date DESC;


'- "daily_revenue": User wants to know the total revenue for a specific day.\n- "current_month_revenue": User wants to know total sales grouped by category for a specific month.\n- "top_category": User wants to know total sales grouped by category for a specific month.\n- "top_subcategory": User wants to know total sales grouped by category for a specific month.\n- "top_profit": User wants to know total sales grouped by category for a specific month.\n- "top_quantity_product": User wants to know total sales grouped by category for a specific month.\n- "top_discount": User wants to know total sales grouped by category for a specific month.\n'

In [ ]:
test = list(["a","b","c"])
type(test)

In [1]:
import google.generativeai as genai
import os



genai.configure(api_key="AIzaSyDWflwYIIy9VfA_FC9xL_cHpMEnamq7k-A")

print("Available Models:")
for m in genai.list_models():
    if 'generateContent' in m.supported_generation_methods:
        print(m.name)

C:\Master in Business Informatics\Trimester 3\Project code\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
C:\Users\vucha\AppData\Local\Temp\ipykernel_30616\389474228.py:1: FutureWarning: 

All support for the `google.generativeai` package has ended. It will no longer be receiving 
updates or bug fixes. Please switch to the `google.genai` package as soon as possible.
See README for more details:

https://github.com/google-gemini/deprecated-generative-ai-python/blob/main/README.md

  import google.generativeai as genai


Available Models:
models/gemini-2.5-flash
models/gemini-2.5-pro
models/gemini-2.0-flash
models/gemini-2.0-flash-001
models/gemini-2.0-flash-lite-001
models/gemini-2.0-flash-lite
models/gemini-2.5-flash-preview-tts
models/gemini-2.5-pro-preview-tts
models/gemma-4-26b-a4b-it
models/gemma-4-31b-it
models/gemini-flash-latest
models/gemini-flash-lite-latest
models/gemini-pro-latest
models/gemini-2.5-flash-lite
models/gemini-2.5-flash-image
models/gemini-3-pro-preview
models/gemini-3-flash-preview
models/gemini-3.1-pro-preview
models/gemini-3.1-pro-preview-customtools
models/gemini-3.1-flash-lite-preview
models/gemini-3.1-flash-lite
models/gemini-3-pro-image-preview
models/nano-banana-pro-preview
models/gemini-3.1-flash-image-preview
models/lyria-3-clip-preview
models/lyria-3-pro-preview
models/gemini-3.1-flash-tts-preview
models/gemini-robotics-er-1.5-preview
models/gemini-robotics-er-1.6-preview
models/gemini-2.5-computer-use-preview-10-2025
models/deep-research-max-preview-04-2026
models/

In [5]:


template_descriptions = ""
for key, value in SQL_TEMPLATES.items():
    template_descriptions += f'- "{key}": {value["description"]}\n'

print(template_descriptions)

- "daily_revenue": Calculates the daily total sales revenue trend over a specified date range. Trigger this when the user asks for daily sales, daily revenue, or earnings day-by-day.
- "current_month_revenue": Calculates the total sales revenue aggregated by month over a specified date range. Trigger this when the user asks for monthly revenue trends, monthly sales, or performance month-by-month.
- "top_category": Identifies the highest revenue-generating product categories over a specific date range. Trigger this when the user asks which broad categories are selling the most or generating the highest revenue.
- "top_subcategory": Identifies the highest revenue-generating product sub-categories over a specific date range. Trigger this for a detailed breakdown of best-selling sub-categories or specific item groups.
- "top_profit": Calculates the total net profit aggregated by day over a specific date range. Trigger this when the user asks about profitability, net earnings after costs, o

In [7]:
route_data = {'route': 'template', 'template_name': 'daily_revenue', 'start_date': '2026-05-14', 'end_date': '2026-05-14', 'extra_filters': {}}

template_name = route_data.get("template_name")
start_date = route_data.get("start_date")
end_date = route_data.get("end_date")
extra_filters = route_data.get("extra_filters")

print(template_name)
print(extra_filters)

if extra_filters:
    print("yes")
else:
    print("no")

daily_revenue
{}
no


In [8]:
{'route': 'template', 'template_name': 'daily_revenue', 'start_date': '2026-05-14', 'end_date': '2026-05-14', 'extra_filters': {}}

{'route': 'template',
 'template_name': 'daily_revenue',
 'start_date': '2026-05-14',
 'end_date': '2026-05-14',
 'extra_filters': {}}

In [1]:
from service.db_manager import DatabaseManager

db = DatabaseManager()
raw_data = db.execute_template("daily_revenue", ["2026-05-14", "2026-05-14"])

print(raw_data)


ERROR:root:Error connecting to the database: 2003: Can't connect to MySQL server on '%-.100s:%u' (%s) (Warning: %u format: a real number is required, not str)
ERROR:root:❌ Cannot execute query: Database connection is missing.


None


In [3]:
import pandas as pd
from datetime import datetime

data = [{'order_date': '2025-05-14', 'total_sales': 4182.062011003494}, {'order_date': '2025-05-14', 'total_sales': 4182.062011003494}, {'order_date': '2025-05-13', 'total_sales': 3066.3779554367065}, {'order_date': '2025-05-13', 'total_sales': 3066.3779554367065}, {'order_date': '2025-05-12', 'total_sales': 970.3840036392212}, {'order_date': '2025-05-12', 'total_sales': 970.3840036392212}, {'order_date': '2025-05-11', 'total_sales': 449.4690008163452}, {'order_date': '2025-05-11', 'total_sales': 449.4690008163452}, {'order_date': '2025-05-09', 'total_sales': 1078.2220001220703}, {'order_date': '2025-05-09', 'total_sales': 1078.2220001220703}, {'order_date': '2025-05-08', 'total_sales': 3658.5539054870605}, {'order_date': '2025-05-08', 'total_sales': 3658.5539054870605}, {'order_date': '2025-05-07', 'total_sales': 2549.4680099487305}, {'order_date': '2025-05-07', 'total_sales': 2549.4680099487305}, {'order_date': '2025-05-06', 'total_sales': 3183.3698024749756}, {'order_date': '2025-05-06', 'total_sales': 3183.3698024749756}, {'order_date': '2025-05-05', 'total_sales': 185.1229968070984}, {'order_date': '2025-05-05', 'total_sales': 185.1229968070984}, {'order_date': '2025-05-04', 'total_sales': 1389.4050059318542}, {'order_date': '2025-05-04', 'total_sales': 1389.4050059318542}, {'order_date': '2025-05-03', 'total_sales': 1386.345998764038}, {'order_date': '2025-05-03', 'total_sales': 1386.345998764038}, {'order_date': '2025-05-02', 'total_sales': 399.1099920272827}, {'order_date': '2025-05-02', 'total_sales': 399.1099920272827}, {'order_date': '2025-05-01', 'total_sales': 4108.369936943054}, {'order_date': '2025-05-01', 'total_sales': 4108.369936943054}]

df = pd.DataFrame(data)

print(df)

    order_date  total_sales
0   2025-05-14  4182.062011
1   2025-05-14  4182.062011
2   2025-05-13  3066.377955
3   2025-05-13  3066.377955
4   2025-05-12   970.384004
5   2025-05-12   970.384004
6   2025-05-11   449.469001
7   2025-05-11   449.469001
8   2025-05-09  1078.222000
9   2025-05-09  1078.222000
10  2025-05-08  3658.553905
11  2025-05-08  3658.553905
12  2025-05-07  2549.468010
13  2025-05-07  2549.468010
14  2025-05-06  3183.369802
15  2025-05-06  3183.369802
16  2025-05-05   185.122997
17  2025-05-05   185.122997
18  2025-05-04  1389.405006
19  2025-05-04  1389.405006
20  2025-05-03  1386.345999
21  2025-05-03  1386.345999
22  2025-05-02   399.109992
23  2025-05-02   399.109992
24  2025-05-01  4108.369937
25  2025-05-01  4108.369937


In [4]:
chart = {
    "chart_type": "line",
    "x_col": "order_date",
    "y_col": "total_sales"
}

chart.get("chart_type")

'line'

Daily revenue fluctuated significantly between May 1st and May 14th, starting at **4,108.37** and reaching a peak of **4,182.06**. This variation suggests that income is currently inconsistent, which affects the predictability of daily cash flow. Although the period opened with strong figures, sales experienced a sharp decline toward the end of the first week before climbing back up through the middle of the month. Identifying the causes of these shifts helps in ensuring more stable earnings throughout the month.

*   Investigate the specific drivers behind the **highest revenue peak** and the **lowest daily total** to understand the causes of these shifts. Pinpointing these factors allows for better management of resources during both busy and slow periods.
*   Review the documentation for the **missing date** in the sequence to ensure all sales were recorded correctly. Confirming data accuracy is vital for maintaining a reliable overview of the business's financial progress.